### PR TITLE
Added more documentation to checks

### PR DESCRIPTION
**Pull request by Enkelmann**
I haven't figured out how to escape the `@` character in `dune build @doc` for inclusion into to the makefile. I also need to figure out how to change the build directory for the documentation without causing dune to segfault. But the generated documentation already looks quite nice!

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -28,5 +28,9 @@ uninstall:
 	cd plugins/cwe_checker_type_inference; make uninstall; cd ../..
 	cd plugins/cwe_checker_type_inference_print; make uninstall; cd ../..
 
+documentation:
+	dune build @doc
+	cp -r _build/default/_doc/_html doc/html
+
```
Comment (tbarabosch) [inline Makefile]: remove the _build directory?

File: Makefile [hunk 2]
```diff
@@ -28,5 +28,9 @@ uninstall:
 	cd plugins/cwe_checker_type_inference; make uninstall; cd ../..
 	cd plugins/cwe_checker_type_inference_print; make uninstall; cd ../..
 
+documentation:
```
Comment (tbarabosch) [inline Makefile]: make doc  ?

File: README.md [hunk 3]
```diff
@@ -61,8 +61,9 @@ If you plan to develop cwe_checker, it is recommended to build it using the prov
 -   Sark (latest) for IDA Pro annotations
 -   pytest >= 3.5.1 (for tests)
 -   SCons >= 3.0.5 (for tests)
+-   odoc >= 1.4 (for documentation)
 
-Just run `make all` to compile and register the plugin with BAP. You can run the test suite via `make test`.
+Just run `make all` to compile and register the plugin with BAP. You can run the test suite via `make test`. Documentation can be built via `make documentation`.
```
Comment (tbarabosch) [inline README.md]: make doc?

File: src/checkers/cwe_248.mli [hunk 4]
```diff
@@ -1,10 +1,27 @@
-(** This module implements a check for CWE-248 (Uncaught Exception)
-An uncaught exception may lead to a crash and subsequentially to other unintended behavior.
-See https://cwe.mitre.org/data/definitions/248.html for detailed description.
+(** This module implements a check for CWE-248: Uncaught Exception.
+
+    An uncaught exception may lead to a crash and subsequentially to other unintended behavior.
```
Comment (tbarabosch) [inline src/checkers/cwe_248.mli]: after a crash, the program should be dead. Hence, no other unintended behavior will happen subsequently. 
Ans shouldn't be 'subsequently'?

File: src/checkers/cwe_248.mli [hunk 5]
```diff
@@ -1,10 +1,27 @@
...
+
+    See {: https://cwe.mitre.org/data/definitions/248.html} for a detailed description.
+
+    {1 How the check works}
+
+    We search for exception throws that are reachable in the callgraph without
```
Comment (tbarabosch) [inline src/checkers/cwe_248.mli]: The tool searches

File: src/checkers/cwe_332.mli [hunk 6]
```diff
@@ -1,8 +1,24 @@
...
-would be calling rand without srand. This could lead to predictable random
-numbers and could, for example, weaken crypto functionality.
-See https://cwe.mitre.org/data/definitions/332.html for detailed description. *)
+(** This module implements a check for CWE332: Insufficient Entropy in PRNG.
+
+    This can happen, for instance, if the PRNG is not seeded. A classical example
+    would be calling rand without srand. This could lead to predictable random
+    numbers and could, for example, weaken crypto functionality.
+
+    See {: https://cwe.mitre.org/data/definitions/332.html} for detailed description.
```
Comment (tbarabosch) [inline src/checkers/cwe_332.mli]: for a detailed

File: src/checkers/cwe_332.mli [hunk 7]
```diff
@@ -1,8 +1,24 @@
...
+
+    {1 How the check works}
+
+    We check whether the program calls rand but not srand.
+
+    {1 False Positives}
+
+    - The program could use rand for non-cryptographic purposes.
```
Comment (tbarabosch) [inline src/checkers/cwe_332.mli]: but this would be still an anti-pattern

File: src/checkers/cwe_426.mli [hunk 8]
```diff
@@ -1,13 +1,37 @@
...
-be used to subvert system integrity.  Use the exec(3) family of functions instead, but not execlp(3) or execvp(3).  system() will not, in
-fact, work properly from programs with set-user-ID or set-group-ID privileges on systems on which /bin/sh is bash version 2, since bash 2
-drops privileges on startup.  (Debian uses a modified bash which does not do this when invoked as sh.)"
+(** This module implements a check for CWE-426: Untrusted Search Path.
+
+    Basically, the program searches for critical resources on an untrusted search
+    path that can be adjusted by an adversary. For example, see Nebula Level 1
+    ({: https://exploit-exercises.com/nebula/level01/}).
+
+    According to "man system" the following problems can arise:
```
Comment (tbarabosch) [inline src/checkers/cwe_426.mli]: According to the manual page of *system*()

File: src/checkers/cwe_476.mli [hunk 9]
```diff
@@ -1,19 +1,49 @@
...
+    Functions like malloc() may return NULL values instead of pointers to indicate
+    failed calls. If one tries to access memory through this return value without
+    checking it for being NULL first, this can crash the program.
+
+    See {: https://cwe.mitre.org/data/definitions/476.html} for a detailed description.
+
+    {1 How the check works}
+
+    We search for an execution path where a memory access using the return value of
+    a symbol is happens before the return value is checked through a conditional
```
Comment (tbarabosch) [inline src/checkers/cwe_476.mli]: remove "is"

File: Makefile (hunk 2 above)
Comment (Enkelmann) [reply Makefile]: make doc doesn't work. Or at least I haven't found a way to escape the `@doc` in the build command in a way such that it doesn't collide with make doc.

Comment (tbarabosch) [reply Makefile]: ok, then leave it as it is

File: Makefile (hunk 1 above)
Comment (Enkelmann) [reply Makefile]: Will be done by `make clean`. However, `make clean` does not delete the contents of *doc/html* and I am not sure whether it should or not.

Comment (Enkelmann) [reply Makefile]: `make clean` deletes *doc/html* now.

**Comment by tbarabosch** [comment]
Looks fine, merging :+1:

**Pull request merged**